blueprints/layered-profile.md

```diff
@@ -0,0 +1,347 @@
+      "CDN_FOO"
+    ],
+    "profileDesc": "Mid Cache - Apache Traffic Server, Amiga 123 Machine, CDN Foo",
+    "profileIds": [10, 11, 15],
+    ⋮
+  }
+}
+```
+
+The following table describes the top level `layered_profile` object for servers:
+
+| field           | type                 | optionality | description                                              |
+| ----------------| ---------------------| ----------- | ---------------------------------------------------------|
+| server          | bigint               | required    | the server id associated with a given profile            |
+| layeredProfiles | text                 | required    | the profile names associated with a server               |
+| order           | bigint               | required    | the order in which a profile is applied to a server      |
+
+**API constraints**
+- In API 5.0, GET /servers object profile and profileId fields to be arrays, but a PUT or POST with multiple values returns a 400 error. 
+  The UI may or may not display a list (which can only add 1), the client implements handling multiple, and the API is documented to potentially return multiple and how their Parameters must be applied.
+- In API 6.0 (or possibly 5.0 in the following TC major version wherein API 4.0 is removed), this feature is actually implemented, and multiple profiles can be assigned, are displayed in the UI, etc.
+
+The only disadvantage to splitting it across mulitple version is a little delay to get the feature deployed.
+But the advantage is that it solves both concerns: 
+- clients on supported APIs are never given 4xx errors even after the feature is in-use
+- we never return or apply partial data that could produce wrong results in production
+
+- Add `Profiles` key to API endpoints for Server objects (and ProfileNames, ProfileIDs)
+
+#### Client Impact
+- Existing Go client methods will be updated for the `/servers` endpoints in order to write TO API tests for the exising endpoints.
+- All functions which get Parameters on Servers must be changed to get the Parameters of all assigned Profiles in order.
+
+#### Data Model / Database Impact
+
+- A new Database Table `server_profiles` as described below, will be created:
+```text
+            Table "traffic_ops.server_profiles"
+     Column    |  Type                    | Collation | Nullable | Default
+---------------+--------------------------+-----------+----------+--------
+ server        | bigint                   |           | not null |
+ profile_names | text                     |           | not null |
+ order         | bigint                   |           | not null |
+Indexes:
+    "pk_server_profile" PRIMARY KEY(profile_names, server, order)
+Foreign-key constraints:
+    "fk_server" FOREIGN KEY (server) REFERENCES server(id)
+    "fk_server" FOREIGN KEY (profile_names) REFERENCES profile(name)
+```
+
+All profiles assigned to a given server will have the same values of:
+- type
+- cdn
+- routing_disabled
```

Comment:
This one sounds hard to enforce. Imagine I have a server that has a bunch of Profiles that are all `routingDisabled` set to `true`. What if I change one to `false`? Does Profile update validation prevent that? If it doesn't, my cache server cannot be updated unless I revert it. If it does, then I need to remove all Profiles from the server, update them all, and reassign in order to make the change. Both are cumbersome. I would say don't bother.

Reply:
Brennan explained above comment over chat.